### ISSUE TITLE
STM orElse method results in partially-rolled-back commits

### ISSUE THREAD

**Issue by dccorona** [scalaz/scalaz-zio#868]
I believe I have found an issue in STM pertaining to the use of `STM.orElse` that results in partially-rolled-back commits. From exploring in a debugger, I'm pretty sure the root cause of the issue is [here](https://github.com/scalaz/scalaz-zio/blob/master/core/shared/src/main/scala/scalaz/zio/stm/STM.scala#L235), where, when an STM fails and triggers its `orElse` block, the *entire* current state of the journal is reset, rather than just resetting the portions of the journal that were modified by the failed STM. 

Am I misunderstanding the purpose of the `orElse` method? In its current form it seems like it is only usable as a top-level "this or that" for use just before actually executing a transaction, rather than a mechanism for graceful failure handling within composed transactions like I thought.

Here's a fairly minimal example that reproduces the issue:

```
val runtime = new DefaultRuntime {}

val ref1 = runtime.unsafeRun(TRef.makeCommit(0))
val ref2 = runtime.unsafeRun(TRef.makeCommit("msg"))

val action1 = STM.atomically {
  for {
    _ <- ref1.set(2)
    newVal <- ref1.get
    computedMsg <- STM.partial(throw new RuntimeException()).orElse(STM.succeed("newMsg"))
    _ <- ref2.set(computedMsg)
    newMsg <- ref2.get
  } yield (newVal, newMsg)
}

val a1res = runtime.unsafeRun(action1)
println(s"Result of running action 1: $a1res")

val action2 = STM.atomically {
  for {
    val1 <- ref1.get
    val2 <- ref2.get
  } yield (val1, val2)
}

val a2res = runtime.unsafeRun(action2)
println(s"Result of running action 2: $a2res")
```

Result of the above is:

```
Result of running action 1: (2,newMsg)
Result of running action 2: (0,newMsg)
```

However, with a small tweak to replace `orElse` with a try/catch, you get correct results:

```
val runtime = new DefaultRuntime {}

val ref1 = runtime.unsafeRun(TRef.makeCommit(0))
val ref2 = runtime.unsafeRun(TRef.makeCommit("msg"))

val action1 = STM.atomically {
  for {
    _ <- ref1.set(2)
    newVal <- ref1.get
    computedMsg <- STM.succeed(try {
      throw new RuntimeException()
    } catch {
      case _: Exception => "newMsg"
    })
    _ <- ref2.set(computedMsg)
    newMsg <- ref2.get
  } yield (newVal, newMsg)
}

val a1res = runtime.unsafeRun(action1)
println(s"Result of running action 1: $a1res")

val action2 = STM.atomically {
  for {
    val1 <- ref1.get
    val2 <- ref2.get
  } yield (val1, val2)
}

val a2res = runtime.unsafeRun(action2)
println(s"Result of running action 2: $a2res")
```
Result of the above is:

```
Result of running action 1: (2,newMsg)
Result of running action 2: (2,newMsg)
```

Which is the result I would expect from the above snippet.

**Comment by jdegoes** [comment]
@dccorona You are right, this is a bug and will be fixed. `orElse` should have "local" semantics, not global semantics.

Thank you for reporting and minimizing! 🙏

**Comment by jdegoes** [comment]
@dccorona Will be fixed in #873
